ui/src/shared/apis/queryCache.ts

```diff
@@ -0,0 +1,164 @@
+// Libraries
+import {sortBy} from 'lodash'
+
+// Utils
+import {asAssignment, getAllVariables} from 'src/variables/selectors'
+import {buildVarsOption} from 'src/variables/utils/buildVarsOption'
+import {filterUnusedVarsBasedOnQuery} from 'src/shared/utils/filterUnusedVars'
+import {runQuery} from 'src/shared/apis/query'
+
+// Types
+import {RunQueryResult} from 'src/shared/apis/query'
+import {CancelBox} from 'src/types/promises'
+import {AppState, Variable} from 'src/types'
+
+export const TIME_INVALIDATION = 5000
+
+const asSimplyKeyValueVariables = (vari: Variable) => {
+  return {
+    [vari.name]: vari.selected || [],
+  }
+}
+
+// Hashing function found here:
+// https://jsperf.com/hashcodelordvlad
+// Through this thread:
+// https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript
+const hashCode = (queryText: string): string => {
+  let hash = 0,
+    char
+  if (!queryText) {
+    return `${hash}`
+  }
+  for (let i = 0; i < queryText.length; i++) {
+    char = queryText.charCodeAt(i)
+    hash = (hash << 5) - hash + char
+    hash |= 0 // Convert to 32bit integer
+  }
+  return `${hash}`
+}
+
+class QueryCache {
+  cache = {}
+
+  _cleanExpiredQueries = (): void => {
+    const now = Date.now()
+    for (const id in this.cache) {
+      if (this.cache[id].isCustomTime) {
+        continue
```

Comment:
i'm confused about this logic - why do we not expire the cache if it's a custom time?

Reply:
We do but we will want to treat that data differently since the data will never change. So instead of clearing out the data after 5s we may want to invalidate it after 10s